paimon-core/src/main/java/org/apache/paimon/mergetree/compact/ChangelogMergeTreeRewriter.java

```diff
@@ -132,8 +136,9 @@ private CompactResult rewriteChangelogCompaction(
 
             while (iterator.hasNext()) {
                 ChangelogResult result = iterator.next();
-                if (rewriteCompactFile && result.result() != null) {
-                    compactFileWriter.write(result.result());
+                KeyValue keyValue = result.result();
+                if (rewriteCompactFile && keyValue != null && (!dropDelete || keyValue.isAdd())) {
```

Comment:
what about also wrap the `ConcatRecordReader.create(sectionReaders)` in `DropDeleteReader` when `dropDelete` is needed?

Reply:
No, this mode we need read delete

Reply:
Got it